search_params.go

```diff
@@ -0,0 +1,104 @@
+package faiss
+
+/*
+#include <faiss/c_api/Index_c.h>
+#include <faiss/c_api/IndexIVF_c.h>
+#include <faiss/c_api/impl/AuxIndexStructures_c.h>
+*/
+import "C"
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+)
+
+type SearchParams struct {
+	sp *C.FaissSearchParameters
+}
+
+// Delete frees the memory associated with s.
+func (s *SearchParams) Delete() {
+	if s == nil || s.sp == nil {
+		return
+	}
+	C.faiss_SearchParameters_free(s.sp)
+}
+
+type searchParamsIVF struct {
+	NprobePct   float32 `json:"ivf_nprobe_pct,omitempty"`
+	MaxCodesPct float32 `json:"ivf_max_codes_pct,omitempty"`
+}
+
+func (s *searchParamsIVF) Validate() error {
+	if s.NprobePct < 0 || s.NprobePct > 100 {
+		return fmt.Errorf("invalid IVF search params, ivf_nprobe_pct:%v, should be in range [0, 100]", s.NprobePct)
+	}
+
+	if s.MaxCodesPct < 0 || s.MaxCodesPct > 100 {
+		return fmt.Errorf("invalid IVF search params, ivf_max_codes_pct:%v, should be in range [0, 100]", s.MaxCodesPct)
+	}
+
+	return nil
+}
+
+// Attempt to create IVF search params
+// and fall back to default search params if the index is not IVF.
+//
+// Always return a valid SearchParams object,
+// thus caller must clean up the object
+// by invoking Delete() method, even if an error is returned.
+func NewSearchParamsIVF(idx Index, params json.RawMessage, sel *C.FaissIDSelector,
+) (*SearchParams, error) {
+
+	rv := &SearchParams{}
+	if c := C.faiss_SearchParameters_new(&rv.sp, sel); c != 0 {
+		return rv, fmt.Errorf("failed to create faiss search params")
+	}
+
+	// # check if the index is IVF and set the search params
+
+	i := C.faiss_IndexIVF_cast(idx.cPtr())
+	if i == nil { // not an IVF index, return the default search params
+		return rv, nil
+	}
+
+	rv.sp = C.faiss_SearchParametersIVF_cast(rv.sp)
+	if len(params) == 0 {
+		return rv, nil
+	}
+
+	var ivf searchParamsIVF
```

Comment:
nit: maybe to be a bit more specific and clearer, this can be renamed to ivfParams?